src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs

```diff
@@ -93,53 +97,125 @@ public sealed class CosmosFhirDataStore : IFhirDataStore, IProvideCapability
             EnsureArg.IsNotNull(resource, nameof(resource));
 
             var cosmosWrapper = new FhirCosmosResourceWrapper(resource);
+            var partitionKey = new PartitionKey(cosmosWrapper.PartitionKey);
 
-            try
-            {
-                _logger.LogDebug($"Upserting {resource.ResourceTypeName}/{resource.ResourceId}, ETag: \"{weakETag?.VersionId}\", AllowCreate: {allowCreate}, KeepHistory: {keepHistory}");
+            _logger.LogDebug($"Upserting {resource.ResourceTypeName}/{resource.ResourceId}, ETag: \"{weakETag?.VersionId}\", AllowCreate: {allowCreate}, KeepHistory: {keepHistory}");
 
-                UpsertWithHistoryModel response = await _retryExceptionPolicyFactory.GetRetryPolicy().ExecuteAsync(
-                    async ct => await _upsertWithHistoryProc.Execute(
-                        _containerScope.Value.Scripts,
-                        cosmosWrapper,
-                        weakETag?.VersionId,
-                        allowCreate,
-                        keepHistory,
-                        ct),
-                    cancellationToken);
-
-                return new UpsertOutcome(response.Wrapper, response.OutcomeType);
+            if (weakETag == null && allowCreate && !cosmosWrapper.IsDeleted)
+            {
+                // Optimistically try to create this as a new resource
+                try
+                {
+                    await _retryExceptionPolicyFactory.GetRetryPolicy().ExecuteAsync(
+                        async ct => await _containerScope.Value.CreateItemAsync(
+                            cosmosWrapper,
+                            partitionKey,
+                            cancellationToken: ct,
+                            requestOptions: new ItemRequestOptions { EnableContentResponseOnWrite = false }),
+                        cancellationToken);
+
+                    return new UpsertOutcome(cosmosWrapper, SaveOutcomeType.Created);
```

Comment:
I'm sure its in there somewhere, where does the Version get set now?

Reply:
[Here](https://github.com/microsoft/fhir-server/blob/64b665a561caa15420c558c2614e3d0d28f2ba54/src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs#L169-L169) on the wrapper
[Here](https://github.com/microsoft/fhir-server/blob/a70c3c745d0d01975ae113dedb7a15b3665bacb5/src/Microsoft.Health.Fhir.Core/Extensions/RawResourceElementExtensions.cs#L65-L69) in the raw resource